In [759]:
import numpy as np
import scipy as sp

Стандартное отклонение

In [760]:
x = [1, 3, 5, 5, 100]
y = [1, 2, 3, 4, 5, 6]

In [761]:
np.std(x) # смещенная оценка

np.float64(38.62848689762518)

In [762]:
np.std(x, ddof=1) # несмещенная оценка

np.float64(43.18796128552493)

Дисперсия


In [763]:
np.var(x) # смещенная оценка

np.float64(1492.16)

In [764]:
np.var(x, ddof=1) # несмещенная оценка

np.float64(1865.2)

Медиана

In [765]:
np.median(x)

np.float64(5.0)

In [766]:
np.median(y)

np.float64(3.5)

Мода

In [767]:
sp.stats.mode(x)

ModeResult(mode=np.int64(5), count=np.int64(2))

Квартили распределения

In [768]:
np.quantile(x, q=0.25)

np.float64(3.0)

In [769]:
z = np.array([1, 2, 4, 2, 1, 5, 7, 2, 3, 5, 7, 8, 9])
z

array([1, 2, 4, 2, 1, 5, 7, 2, 3, 5, 7, 8, 9])

In [770]:
n = len(z)
n

13

In [771]:
z.sort()
z

array([1, 1, 2, 2, 2, 3, 4, 5, 5, 7, 7, 8, 9])

In [772]:
# Если n*k/100 - целое число, то k-я персентиль это среднее значений под номерами n*k/100 и n*k/100 + 1

In [773]:
# Если n*k/100 - не целое число, то k-я персентиль совпадает с измерениями j+1, где j - максимальное целое число < n*k/100

In [774]:
# посчитаем 25ю персентиль

In [775]:
k = 25

In [776]:
n * k/100

3.25

In [777]:
j = 3

In [778]:
# 25 персентиль соответствует j+1=4, т.е. 4-му значению, которое равно

In [779]:
z[3]

np.int64(2)

In [780]:
k = 75

In [781]:
n * k / 100

9.75

In [782]:
z[9]

np.int64(7)

1. Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.

In [783]:
x = np.array([100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150])

среднее

In [784]:
x_sr = sum(x) / len(x)
x_sr

np.float64(65.3)

std

In [785]:
y = (x - x_sr)**2
y

array([1.20409e+03, 2.16090e+02, 9.40900e+01, 1.36890e+02, 5.61690e+02,
       1.04329e+03, 4.12090e+02, 1.62409e+03, 9.00000e-02, 2.33289e+03,
       1.24609e+03, 1.70569e+03, 6.88900e+01, 1.06090e+02, 2.20900e+01,
       9.40900e+01, 9.00000e-02, 3.49690e+02, 6.10090e+02, 7.17409e+03])

In [786]:
S = (sum(y) / (len(y) - 1))**0.5 # несмещенное стандартное квадратичное отклонение
S

np.float64(31.624607341019814)

In [787]:
np.std(x, ddof=1) # несмещенное стандартное квадратичное отклонение

np.float64(31.624607341019814)

дисперсия

In [788]:
D = S**2 # несмещенная оценка дисперсии
D

np.float64(1000.1157894736843)

In [789]:
np.var(x, ddof=1) # несмещенная оценка дисперсии

np.float64(1000.1157894736842)

In [790]:
S  = (sum(y) / (len(y)))**0.5 # смещенное стандартное квадратичное отклонение
S

np.float64(30.823854398825596)

In [791]:
D = S**2 # смещенная оценка дисперсии
D

np.float64(950.11)

In [792]:
np.var(x) # смещенная оценка дисперсии

np.float64(950.11)

In [793]:
x = [202, 203, 199, 197, 195, 201, 200, 204, 194, 190] 

In [794]:
np.std(x, ddof=1)

np.float64(4.453463071962462)

In [795]:
np.mean(x)

np.float64(198.5)

In [796]:
np.std(x, ddof=1)


np.float64(4.453463071962462)

коэффициент корреляции

In [797]:
s = np.array([27, 37, 42, 48, 54, 56, 77, 80])

In [798]:
p = np.array([1.2, 1.6, 1.8, 1.8, 2.5, 2.6, 3, 3.3])

In [799]:
np.corrcoef(s, p)

array([[1.        , 0.97318772],
       [0.97318772, 1.        ]])

In [800]:
np.mean(p)

np.float64(2.2249999999999996)

In [801]:
np.random.normal(size=20)

array([-0.50856773, -0.43534508,  0.07383893, -0.20400298, -0.91857238,
        0.86346673, -0.0790598 , -0.20896425,  1.92505922, -1.31066955,
       -0.19631382,  0.78169666, -0.24618263,  0.43816846,  0.91701305,
        0.19985064, -0.64139927,  1.19263923,  0.87373302,  0.85079178])

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

In [802]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])

In [803]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [804]:
both = zp * ks
both

array([ 14035,  25830, 166060, 183800,  18360,  51730,  35262, 135300,
        89520,  91520])

In [805]:
cov = np.mean(both) - np.mean(zp) * np.mean(ks)
cov

np.float64(9157.839999999997)

In [806]:
std_zp = np.std(zp)
std_zp

np.float64(59.115480206118605)

In [807]:
std_ks = np.std(ks)
std_ks

np.float64(174.55340157098058)

In [808]:
r = cov / (std_ks * std_zp)
r

np.float64(0.8874900920739158)

In [809]:
np.corrcoef(zp, ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

In [810]:
a = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
np.mean(a)

np.float64(118.1)

In [811]:
np.std(a, ddof=1)

np.float64(10.54566788359614)

# Метод градиентного спуска (минимизация модели)

## без интерсепта

In [812]:
X = np.array([27, 37, 42, 48, 54, 56, 77, 80])
X

array([27, 37, 42, 48, 54, 56, 77, 80])

In [813]:
y = np.array([1.2, 1.6, 1.8, 1.8, 2.5, 2.6, 3, 3.3])
y

array([1.2, 1.6, 1.8, 1.8, 2.5, 2.6, 3. , 3.3])

## mse = (1/n) * sum(y - b1 * X)^2

## mse_p = (2/n) * sum((y - b1 * X) * (-X)) - производная от mse, где B1 это переменная, а x и y это константы

In [814]:
def mse_(B1, y=y, X=X, n=8):
    return np.sum((y - B1 * X)**2) / n

In [815]:
alpha = 1e-6 # скорость обучения (спуск вниз на 1e-6)
alpha

1e-06

In [816]:
B1 = 0.1 # берем из стандартного нормального распределения [-3; 3] (99,7%)

In [817]:
n = 8

In [818]:
for i in range(10):
    B1 -= alpha * (2 / n) * np.sum((y - B1*X) * (-X))
    print(f'B1={B1}')

B1=0.099643625
B1=0.09928943698428125
B1=0.09893742253186796
B1=0.09858756830414553
B1=0.09823986104435506
B1=0.09789428757709111
B1=0.0975508348078024
B1=0.09720948972229562
B1=0.09687023938624233
B1=0.0965330709446888


In [819]:
for i in range(3000):
    B1 -= alpha * (2 / n) * np.sum((y - B1*X) * (-X))
    if i % 10 == 0:
        print(f'Iteration {i}: B1={B1}, mse={mse_(B1)}')

Iteration 0: B1=0.09619797162156898, mse=9.065656785336
Iteration 10: B1=0.09295802461401302, mse=8.018824950239837
Iteration 20: B1=0.08991150325539897, mse=7.093254327129224
Iteration 30: B1=0.08704685998597309, mse=6.27489845597186
Iteration 40: B1=0.08435323663819518, mse=5.551337967840769
Iteration 50: B1=0.08182042327985041, mse=4.9115921085679695
Iteration 60: B1=0.07943881951423788, mse=4.345952094816223
Iteration 70: B1=0.07719939809074752, mse=3.845833773580402
Iteration 80: B1=0.07509367068789481, mse=3.403647349079136
Iteration 90: B1=0.0731136557391167, mse=3.0126822000119056
Iteration 100: B1=0.07125184817937577, mse=2.6670050391683473
Iteration 110: B1=0.06950119099789967, mse=2.361369869859907
Iteration 120: B1=0.06785504848922905, mse=2.0911383726726376
Iteration 130: B1=0.06630718110118475, mse=1.8522095143304953
Iteration 140: B1=0.06485172178441763, mse=1.6409573104132085
Iteration 150: B1=0.06348315375389614, mse=1.4541757974157041
Iteration 160: B1=0.0621962895780

Видно что последние значения уже меняются больше чем 10 знаков после запятой, можем взять значения

y(штрих) = 0.042x

In [820]:
mse_(0.041927731027279896)

np.float64(0.02849747423310478)

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

In [821]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
zp

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [822]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
ks

array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

без интерсепта

In [823]:
y = ks.reshape((10, 1))
y

array([[401],
       [574],
       [874],
       [919],
       [459],
       [739],
       [653],
       [902],
       [746],
       [832]])

In [824]:
x = zp.reshape((10, 1))
x

array([[ 35],
       [ 45],
       [190],
       [200],
       [ 40],
       [ 70],
       [ 54],
       [150],
       [120],
       [110]])

In [825]:
x.T @ y

array([[811417]])

In [826]:
np.linalg.inv(np.dot(x.T, x))

array([[7.25868502e-06]])

In [827]:
B = np.dot(np.linalg.inv(np.dot(x.T, x)), x.T @ y) # значения B1, без интерсепта
B

array([[5.88982042]])

с интерсептом

In [828]:
ones_matrix = np.ones((10, 1))
ones_matrix

array([[1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.]])

In [829]:
x = np.hstack((ones_matrix, x))
x

array([[  1.,  35.],
       [  1.,  45.],
       [  1., 190.],
       [  1., 200.],
       [  1.,  40.],
       [  1.,  70.],
       [  1.,  54.],
       [  1., 150.],
       [  1., 120.],
       [  1., 110.]])

In [830]:
B = np.dot(np.linalg.inv(np.dot(x.T, x)), x.T @ y)
B

array([[444.17735732],
       [  2.62053888]])

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [831]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
zp

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [832]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
ks

array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [833]:
def mse_(B1, y=ks, x=zp, n=10):
    return np.sum((y - B1 * x)**2) / n

In [834]:
alpha = 1e-6
alpha

1e-06

In [835]:
B1 = 0.1

In [836]:
n = 10

In [837]:
for i in range(3000):
    B1 -= alpha * (2 / n) * np.sum((ks - B1*zp) * (-zp))
    if i % 10 == 0:
        print(f'Iteration {i}: B1={B1}, mse={mse_(B1)}')

Iteration 0: B1=0.25952808, mse=493237.7212546963
Iteration 10: B1=1.6319792141937546, mse=306275.7568040035
Iteration 20: B1=2.6698789606516935, mse=199352.96411174876
Iteration 30: B1=3.4547782236263824, mse=138204.25062502263
Iteration 40: B1=4.04834889855975, mse=103233.54824312925
Iteration 50: B1=4.497229618367758, mse=83233.94472982832
Iteration 60: B1=4.836690291080364, mse=71796.25222021657
Iteration 70: B1=5.093403362579137, mse=65255.0820486468
Iteration 80: B1=5.287539550879594, mse=61514.21323463749
Iteration 90: B1=5.434352720103083, mse=59374.825462025976
Iteration 100: B1=5.54537842245223, mse=58151.31823171113
Iteration 110: B1=5.629340281237233, mse=57451.59938606899
Iteration 120: B1=5.692835432193034, mse=57051.433003054815
Iteration 130: B1=5.740852880272922, mse=56822.57946410561
Iteration 140: B1=5.777165500770666, mse=56691.69904907657
Iteration 150: B1=5.804626485478126, mse=56616.849068093856
Iteration 160: B1=5.825393526544474, mse=56574.04266614384
Iteration

получилось что B1 = 5.8898 (это все без интерсепта B0)

Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться
на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).

In [838]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
zp

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [839]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
ks

array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [840]:
def mse_(B0, B1, y=ks, x=zp, n=10):
    return np.sum((y - (B0 + B1 * x))**2) / n

In [841]:
alpha = 5e-5
alpha

5e-05

In [842]:
B0 = 0.1
B1 = 0.1

In [843]:
n = 10

In [844]:
for i in range(1000000):
    B0 += alpha * (2 / n) * np.sum((ks - (B0 + B1 * zp)))
    B1 -= alpha * (2 / n) * np.sum((ks - (B0 + B1 * zp)) * (-zp))
    if i % 10000 == 0:
        print(f'Iteration {i}: B0={B0},B1={B1}, mse={mse_(B0, B1)}')

Iteration 0: B0=0.169966,B1=8.074680544760001, mse=122318.06397097581
Iteration 10000: B0=99.63944588565495,B1=5.156425913533925, mse=36582.05548914076
Iteration 20000: B0=176.83480207766843,B1=4.588248374483324, mse=24600.363270732934
Iteration 30000: B0=236.7341719071708,B1=4.147373712855302, mse=17386.293881109214
Iteration 40000: B0=283.2128033036884,B1=3.8052791140606974, mse=13042.767423439807
Iteration 50000: B0=319.27767648420047,B1=3.5398324356503275, mse=10427.569111705801
Iteration 60000: B0=347.2620434598509,B1=3.333860346855704, mse=8852.981670752433
Iteration 70000: B0=368.9763800418267,B1=3.1740372832691417, mse=7904.936699140194
Iteration 80000: B0=385.8255163330167,B1=3.0500233365073157, mse=7334.127327222927
Iteration 90000: B0=398.89952357644796,B1=2.9537953037921687, mse=6990.448125471097
Iteration 100000: B0=409.0442373734796,B1=2.879127619051743, mse=6783.521961452364
Iteration 110000: B0=416.91597983984525,B1=2.8211895838744083, mse=6658.933574557688
Iteration 12

### ANOVA (однофакторный диспирсионный анализ)

In [845]:
from scipy import stats

In [846]:
stats.f_oneway?

Signature: stats.f_oneway(*samples, axis=0, nan_policy='propagate', keepdims=False)
Docstring:
Perform one-way ANOVA.

The one-way ANOVA tests the null hypothesis that two or more groups have
the same population mean.  The test is applied to samples from two or
more groups, possibly with differing sizes.

Parameters
----------
sample1, sample2, ... : array_like
    The sample measurements for each group.  There must be at least
    two arguments.  If the arrays are multidimensional, then all the
    dimensions of the array must be the same except for `axis`.
axis : int or None, default: 0
    If an int, the axis of the input along which to compute the statistic.
    The statistic of each axis-slice (e.g. row) of the input will appear in a
    corresponding element of the output.
    If ``None``, the input will be raveled before computing the statistic.
nan_policy : {'propagate', 'omit', 'raise'}
    Defines how to handle input NaNs.
    
    - ``propagate``: if a NaN is present in the 

In [847]:
y1 = np.array([70, 50, 65, 60, 75])
y1

array([70, 50, 65, 60, 75])

In [848]:
y2 = np.array([80, 75, 90, 70, 75, 65, 85, 100])
y2

array([ 80,  75,  90,  70,  75,  65,  85, 100])

In [849]:
y3 = np.array([130, 100, 140, 150, 160, 170, 200])
y3

array([130, 100, 140, 150, 160, 170, 200])

In [850]:
stats.f_oneway(y1, y2, y3)

F_onewayResult(statistic=np.float64(31.77269601100413), pvalue=np.float64(1.8091304567650962e-06))

### 2-WAY ANOVA

In [851]:
y111 = 57

In [852]:
y112 = 59

In [853]:
y11 = (y111 + y112) / 2
y11

58.0

In [854]:
y121 = 56

In [855]:
y122 = 58

In [856]:
y12 = (y121 + y122) / 2
y12

57.0

In [857]:
y211 = 32

In [858]:
y212 = 34

In [859]:
y21 = (y211 + y212) / 2
y21

33.0

In [860]:
y221 = 71

In [861]:
y222 = 71

In [862]:
y22 = (y221 + y222) / 2
y22

71.0

In [863]:
YcpA1 = (y11 + y12) / 2
YcpA1

57.5

In [864]:
YcpA2 = (y21 + y22) / 2
YcpA2

52.0

In [865]:
YcpB1 = (y11 + y21) / 2
YcpB1

45.5

In [866]:
YcpB2 = (y12 + y22) / 2
YcpB2

64.0

In [867]:
Ycp = (YcpA1 + YcpA2 + YcpB1 + YcpB2) / 4
Ycp

54.75

дальше все в тетради написано

#### Нормальный метод

In [868]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
import pandas as pd

In [ ]:
fa = ['low', 'low', 'low', 'low', 'high', 'high', 'high', 'high'] # уровни

In [ ]:
fb = ['low', 'low', 'high', 'high', 'low', 'low', 'high', 'high'] # уровни

In [874]:
treatments = [57, 59, 56, 58, 32, 34, 71, 71]

In [881]:
df = pd.DataFrame({'fA': fa, 'fb': fb, 'treatments': treatments})
df

fA    fb  treatments
0   low   low          57
1   low   low          59
2   low  high          56
3   low  high          58
4  high   low          32
5  high   low          34
6  high  high          71
7  high  high          71

In [883]:
lm_model = ols('treatments ~ C(fa) * C(fb)', data=df).fit()

In [884]:
table = sm.stats.anova_lm(lm_model, typ=2)
table

sum_sq   df           F    PR(>F)
C(fa)          60.5  1.0   40.333333  0.003150
C(fb)         684.5  1.0  456.333333  0.000028
C(fa):C(fb)   760.5  1.0  507.000000  0.000023
Residual        6.0  4.0         NaN       NaN

Везде принимаем альтернативную гипотезу

Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов. Даны значения роста в трех группах случайно выбранных спортсменов: Футболисты: 173, 175, 180, 178, 177, 185, 183, 182. Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180. Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.

In [885]:
a = [173, 175, 180, 178, 177, 185, 183, 182]

In [886]:
b = [177, 179, 180, 188, 177, 172, 171, 184, 180]

In [887]:
c = [172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170]

In [890]:
sp.stats.f_oneway(a, b, c)

F_onewayResult(statistic=np.float64(5.500053450812596), pvalue=np.float64(0.010482206918698694))